File: workflows/org.linxira.visualization-ggplot2/src/render.R
#!/usr/bin/env Rscript
# Deterministic ggplot2 renderer for Linxira PlotSpec (M1-T2).
#
# Same request/result contract as the matplotlib pack: request JSON
# {"plot_spec": ..., "output_path": "<relative>"} renders the figure and
# writes a render-result JSON whose data_summary (point counts, coordinate
# ranges) must match the matplotlib backend for the same spec. svglite and
# cairo png produce timestamp-free bytes, so identical specs are
# byte-identical; base pdf embeds a creation date, which is disclosed in
# warnings instead of hidden.

suppressPackageStartupMessages({
  library(jsonlite)
  library(ggplot2)
  library(svglite)
  library(digest)
})

schema_version <- "1"
backend_name <- "ggplot2"

canonical_hash <- function(value) {
  digest_json <- jsonlite::toJSON(value, auto_unbox = TRUE, digits = NA, null = "null")
  digest::digest(as.character(digest_json), algo = "sha256", serialize = FALSE)
}

finite <- function(values) {
  values[vapply(values, function(v) is.finite(suppressWarnings(as.numeric(v))), logical(1))]
}

numeric_or_null <- function(x) if (length(x) == 0) NULL else as.numeric(x)

data_summary <- function(data) {
  kind <- data$kind %||% ""
  if (kind %in% c("scatter", "line", "bar")) {
    series <- data$series %||% list()
    xs <- numeric(0); ys <- numeric(0); counts <- integer(0)
    for (entry in series) {
      xs <- c(xs, finite(entry$x %||% list()))
      ys <- c(ys, finite(entry$y %||% list()))
      counts <- c(counts, min(length(finite(entry$x %||% list())), length(finite(entry$y %||% list()))))
    }
    return(list(kind = kind, series_count = length(series),
                point_counts = as.integer(counts),
                x_range = if (length(xs)) range(xs) else NULL,
                y_range = if (length(ys)) range(ys) else NULL))
  }
  if (kind %in% c("box", "violin")) {
    groups <- data$groups %||% list()
    values <- numeric(0); counts <- integer(0)
    for (group in groups) {
      group_values <- finite(group$values %||% list())
      values <- c(values, group_values)
      counts <- c(counts, length(group_values))
    }
    return(list(kind = kind, series_count = length(groups),
                point_counts = as.integer(counts),
                x_range = NULL,
                y_range = if (length(values)) range(values) else NULL))
  }
  if (identical(kind, "heatmap")) {
    matrix_values <- data$matrix$values
    flat <- finite(unlist(matrix_values, use.names = FALSE))
    return(list(kind = kind, series_count = 1L,
                point_counts = nrow(matrix_values) * ncol(matrix_values),
                x_range = NULL,
                y_range = if (length(flat)) range(flat) else NULL))
  }
  stop(sprintf("unsupported plot kind '%s'", kind))
}

build_frame <- function(data) {
  kind <- data$kind %||% ""
  if (kind %in% c("scatter", "line", "bar")) {
    rows <- list()
    for (index in seq_along(data$series %||% list())) {
      entry <- data$series[[index]]
      rows[[length(rows) + 1]] <- data.frame(
        x = as.character(unlist(entry$x %||% list(), use.names = FALSE)),
        x_num = suppressWarnings(as.numeric(unlist(entry$x %||% list(), use.names = FALSE))),
        y = as.numeric(unlist(entry$y %||% list(), use.names = FALSE)),
        series = entry$name %||% sprintf("series %d", index)
      )
    }
    return(do.call(rbind, rows))
  }
  if (kind %in% c("box", "violin")) {
    rows <- list()
    for (index in seq_along(data$groups %||% list())) {
      entry <- data$groups[[index]]
      rows[[length(rows) + 1]] <- data.frame(
        group = entry$name %||% sprintf("group %d", index),
        value = as.numeric(finite(entry$values %||% list()))
      )
    }
    return(do.call(rbind, rows))
  }
  NULL
}

apply_theme <- function(plot, spec) {
  theme_name <- spec$theme %||% "light"
  base <- ggplot2::theme_minimal(base_size = spec$font$size %||% 12)
  if (identical(theme_name, "dark")) {
    base <- base + ggplot2::theme(
      plot.background = ggplot2::element_rect(fill = "#1c1c1c", colour = "#1c1c1c"),
      panel.background = ggplot2::element_rect(fill = "#242424", colour = NA),
      text = ggplot2::element_text(colour = "white"),
      axis.text = ggplot2::element_text(colour = "white")
    )
  } else if (identical(theme_name, "publication")) {
    base <- base + ggplot2::theme(
      panel.grid.minor = ggplot2::element_blank(),
      text = ggplot2::element_text(family = spec$font$family %||% "serif")
    )
  }
  if (!isTRUE(spec$grid %||% TRUE)) {
    base <- base + ggplot2::theme(panel.grid = ggplot2::element_blank())
  }
  plot + base
}

render <- function(spec, output_path) {
  warnings <- character(0)
  data <- spec$data
  summary <- data_summary(data)
  kind <- data$kind %||% ""

  figure <- spec$figure %||% list()
  width <- figure$width %||% 800
  height <- figure$height %||% 600
  dpi <- figure$dpi %||% 150
  palette_name <- spec$palette %||% "set2"
  palette_map <- c(set2 = "Set2", set1 = "Set1", pastel = "Pastel1", vivid = "Set3")
  brewer_name <- palette_map[[tolower(palette_name)]]
  if (is.na(brewer_name)) {
    warnings <- c(warnings, sprintf("unknown palette '%s'; falling back to Set2", palette_name))
    brewer_name <- "Set2"
  }

  frame <- build_frame(data)
  if (kind %in% c("scatter", "line", "bar")) {
    frame$x_num[!is.finite(frame$x_num)] <- NA_real_
    if (identical(kind, "bar")) {
      plot <- ggplot2::ggplot(frame, ggplot2::aes(x = x, y = y, fill = series)) +
        ggplot2::geom_col(position = ggplot2::position_dodge()) +
        ggplot2::scale_fill_brewer(palette = brewer_name)
    } else {
      plot <- ggplot2::ggplot(frame, ggplot2::aes(x = x_num, y = y, colour = series)) +
        ggplot2::scale_colour_brewer(palette = brewer_name)
      if (identical(kind, "line")) {
        plot <- plot + ggplot2::geom_line(linewidth = 0.8)
      } else {
        plot <- plot + ggplot2::geom_point()
      }
    }
    if (isTRUE(spec$legend %||% TRUE)) plot <- plot else plot <- plot + ggplot2::theme(legend.position = "none")
  } else if (kind %in% c("box", "violin")) {
    frame$group <- factor(frame$group, levels = unique(frame$group))
    plot <- ggplot2::ggplot(frame, ggplot2::aes(x = group, y = value, fill = group)) +
      ggplot2::scale_fill_brewer(palette = brewer_name) +
      (if (identical(kind, "violin")) ggplot2::geom_violin() else ggplot2::geom_boxplot()) +
      ggplot2::theme(legend.position = "none")
  } else if (identical(kind, "heatmap")) {
    matrix_values <- matrix(as.numeric(unlist(data$matrix$values, use.names = FALSE)),
                            nrow = length(data$matrix$values), byrow = TRUE)
    long_frame <- expand.grid(row_index = seq_len(nrow(matrix_values)),
                              col_index = seq_len(ncol(matrix_values)))
    long_frame$value <- as.vector(matrix_values)
    plot <- ggplot2::ggplot(long_frame, ggplot2::aes(x = col_index, y = row_index, fill = value)) +
      ggplot2::geom_tile() +
      ggplot2::scale_fill_gradient2(low = "#3b4cc0", mid = "#f7f7f7", high = "#b40426",
                                    midpoint = mean(long_frame$value)) +
      ggplot2::scale_y_reverse() +
      ggplot2::coord_equal()
  }

  title_parts <- c(spec$title %||% "", spec$subtitle %||% "")
  title_parts <- title_parts[nzchar(title_parts)]
  if (length(title_parts)) plot <- plot + ggplot2::ggtitle(title_parts[1])
  x_label <- spec$x_label %||% data$x_label_hint %||% "x"
  y_label <- spec$y_label %||% data$y_label_hint %||% "y"
  plot <- plot + ggplot2::xlab(x_label) + ggplot2::ylab(y_label)
  if (isTRUE(spec$x_log %||% FALSE)) plot <- plot + ggplot2::scale_x_log10()
  if (isTRUE(spec$y_log %||% FALSE)) plot <- plot + ggplot2::scale_y_log10()
  if (!is.null(numeric_or_null(spec$x_range))) plot <- plot + ggplot2::xlim(spec$x_range)
  if (!is.null(numeric_or_null(spec$y_range))) plot <- plot + ggplot2::ylim(spec$y_range)
  plot <- apply_theme(plot, spec)

  fmt <- spec$output$format %||% "svg"
  if (!fmt %in% c("svg", "png", "pdf")) {
    warnings <- c(warnings, sprintf("unsupported format '%s'; rendered svg", fmt))
    fmt <- "svg"
  }
  if (identical(fmt, "pdf")) {
    warnings <- c(warnings, "pdf output embeds a creation date; byte-stable reruns are guaranteed for svg and png only")
  }
  output_path <- sub("\\.[^.]*$", paste0(".", fmt), output_path)
  ggplot2::ggsave(output_path, plot = plot,
                  width = width / dpi, height = height / dpi, dpi = dpi)

  list(
    schema_version = schema_version,
    backend = backend_name,
    plot_spec_sha256 = canonical_hash(spec),
    artifact = list(
      path = basename(output_path),
      format = fmt,
      width = width, height = height, dpi = dpi,
      bytes = as.integer(file.info(output_path)$size)
    ),
    data_summary = summary,
    warnings = warnings
  )
}

main <- function(argv) {
  request_path <- NULL; result_path <- NULL
  i <- 1L
  while (i <= length(argv)) {
    if (identical(argv[[i]], "--request")) request_path <- argv[[i + 1L]]
    if (identical(argv[[i]], "--result")) result_path <- argv[[i + 1L]]
    i <- i + 2L
  }
  if (is.null(request_path) || is.null(result_path)) {
    message("usage: render.R --request REQUEST --result RESULT")
    quit(status = 2L)
  }
  request <- jsonlite::fromJSON(request_path, simplifyVector = FALSE)
  spec <- request$plot_spec
  output_path <- file.path(dirname(request_path), request$output_path %||% "figure.svg")
  result <- tryCatch(render(spec, output_path), error = function(e) {
    list(schema_version = schema_version, backend = backend_name,
         status = "error", message = paste0(class(e)[1], ": ", conditionMessage(e)))
  })
  jsonlite::write_json(result, result_path, auto_unbox = TRUE, null = "null",
                       na = "null", digits = NA, pretty = TRUE)
  if (!is.null(result$status) && identical(result$status, "error")) quit(status = 1L)
  invisible(0L)
}

`%||%` <- function(left, right) if (is.null(left)) right else left

main(commandArgs(trailingOnly = TRUE))

File: workflows/org.linxira.visualization-matplotlib/src/render.py
#!/usr/bin/env python3
"""Deterministic matplotlib renderer for Linxira PlotSpec (M1-T1).

Contract: a request JSON {"plot_spec": {...}, "output_path": "<relative>"}
produces the figure named by output_path (caller-composed filename, any of
svg/png/pdf) plus a render-result JSON with the PlotSpec hash, artifact
metadata, and a data summary (point counts / coordinate ranges) so the
ggplot2 backend can be asserted data-consistent without pixel comparison.

Every PlotSpec knob (title, labels, ranges, log axes, grid, theme, palette,
legend, figure size/dpi, font, format) is honored; defaults follow
schemas/plot-spec.schema.json. Output is deterministic: no timestamps, fixed
dpi, so identical specs render to identical bytes.
"""
from __future__ import annotations

import hashlib
import json
import sys
from pathlib import Path

import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt  # noqa: E402

# SVG clip-path/marker ids get random suffixes unless a salt is pinned; a
# fixed salt is what makes identical specs render to identical bytes.
plt.rcParams["svg.hashsalt"] = "linxira-bio"

SCHEMA_VERSION = "1"
BACKEND = "matplotlib"
KINDS = ("scatter", "line", "bar", "box", "violin", "heatmap")
FORMATS = ("svg", "png", "pdf")
# matplotlib savefig metadata keys that suppress nondeterministic fields.
NO_DATE_METADATA = {
    "svg": {"Date": None, "Creator": None},
    "png": {"Software": None},
    "pdf": {"CreationDate": None, "ModDate": None, "Creator": None},
}


def canonical_hash(value: object) -> str:
    return hashlib.sha256(
        json.dumps(value, sort_keys=True, ensure_ascii=False).encode("utf-8")
    ).hexdigest()


def finite(values: list[float]) -> list[float]:
    return [float(v) for v in values if isinstance(v, (int, float))]


def data_summary(plot_spec: dict) -> dict:
    data = plot_spec["data"]
    kind = data.get("kind", "")
    if kind in ("scatter", "line", "bar"):
        series = data.get("series", [])
        xs = [v for entry in series for v in finite(entry.get("x", []))]
        ys = [v for entry in series for v in finite(entry.get("y", []))]
        return {
            "kind": kind,
            "series_count": len(series),
            "point_counts": [
                min(len(finite(e.get("x", []))), len(finite(e.get("y", []))))
                for e in series
            ],
            "x_range": [min(xs), max(xs)] if xs else None,
            "y_range": [min(ys), max(ys)] if ys else None,
        }
    if kind in ("box", "violin"):
        groups = data.get("groups", [])
        values = [v for group in groups for v in finite(group.get("values", []))]
        return {
            "kind": kind,
            "series_count": len(groups),
            "point_counts": [len(finite(g.get("values", []))) for g in groups],
            "x_range": None,
            "y_range": [min(values), max(values)] if values else None,
        }
    if kind == "heatmap":
        matrix = data["matrix"]
        flat = [v for row in matrix["values"] for v in finite(row)]
        return {
            "kind": kind,
            "series_count": 1,
            "point_counts": [len(matrix["row_labels"]) * len(matrix["col_labels"])],
            "x_range": None,
            "y_range": [min(flat), max(flat)] if flat else None,
        }
    raise ValueError(f"unsupported plot kind {kind!r}; expected one of {KINDS}")


def apply_style(figure, axes, plot_spec: dict, warnings: list[str]) -> None:
    theme = plot_spec.get("theme", "light")
    if theme == "dark":
        figure.patch.set_facecolor("#1c1c1c")
        axes.set_facecolor("#242424")
        axes.tick_params(colors="white")
        for spine in axes.spines.values():
            spine.set_color("white")
        axes.xaxis.label.set_color("white")
        axes.yaxis.label.set_color("white")
        axes.title.set_color("white")
    elif theme == "publication":
        for spine in axes.spines.values():
            spine.set_linewidth(0.8)
        figure.subplots_adjust(left=0.14, right=0.95, top=0.88, bottom=0.12)
    if plot_spec.get("grid", True):
        axes.grid(True, alpha=0.35 if theme != "dark" else 0.25)
    else:
        axes.grid(False)


def palette_colors(plot_spec: dict, count: int, warnings: list[str]):
    name = plot_spec.get("palette", "set2")
    colormap = matplotlib.colormaps.get(name)
    if colormap is None:
        warnings.append(f"unknown palette {name!r}; falling back to Set2")
        colormap = matplotlib.colormaps["Set2"]
    slots = max(count, 1)
    return [colormap(i / max(slots - 1, 1)) for i in range(slots)]


def render(plot_spec: dict, output_path: Path) -> dict:
    warnings: list[str] = []
    data = plot_spec["data"]
    kind = data.get("kind", "")
    summary = data_summary(plot_spec)

    figure_block = plot_spec.get("figure", {}) or {}
    width = int(figure_block.get("width", 800))
    height = int(figure_block.get("height", 600))
    dpi = int(figure_block.get("dpi", 150))
    font_block = plot_spec.get("font", {}) or {}
    font_size = int(font_block.get("size", 12))
    family = font_block.get("family") or None
    if family:
        plt.rcParams["font.family"] = family
    plt.rcParams["font.size"] = font_size

    figure, axes = plt.subplots(figsize=(width / dpi, height / dpi), dpi=dpi)
    if kind in ("scatter", "line"):
        colors = palette_colors(plot_spec, len(data.get("series", [])), warnings)
        for index, entry in enumerate(data.get("series", [])):
            style = {} if kind == "scatter" else {"marker": "o"}
            axes.plot(
                entry.get("x", []),
                entry.get("y", []),
                linestyle="none" if kind == "scatter" else "-",
                color=colors[index],
                label=entry.get("name", f"series {index + 1}"),
                **style,
            )
        if plot_spec.get("legend", True) and data.get("series"):
            axes.legend()
    elif kind == "bar":
        colors = palette_colors(plot_spec, len(data.get("series", [])), warnings)
        for index, entry in enumerate(data.get("series", [])):
            axes.bar(
                [str(x) for x in entry.get("x", [])],
                entry.get("y", []),
                color=colors[index],
                label=entry.get("name", f"series {index + 1}"),
            )
        if plot_spec.get("legend", True) and len(data.get("series", [])) > 1:
            axes.legend()
    elif kind in ("box", "violin"):
        groups = data.get("groups", [])
        colors = palette_colors(plot_spec, len(groups), warnings)
        names = [g.get("name", f"group {i + 1}") for i, g in enumerate(groups)]
        values = [finite(g.get("values", [])) for g in groups]
        if kind == "box":
            axes.boxplot(values, patch_artist=True)
            for patch, color in zip(axes.patches, colors):
                patch.set_facecolor(color)
        else:
            axes.violinplot(values, showmedians=True)
        axes.set_xticks(range(1, len(names) + 1), names)
    elif kind == "heatmap":
        matrix = data["matrix"]
        image = axes.imshow(matrix["values"], aspect="auto", interpolation="nearest")
        figure.colorbar(image, ax=axes)
        axes.set_xticks(range(len(matrix["col_labels"])), matrix["col_labels"])
        axes.set_yticks(range(len(matrix["row_labels"])), matrix["row_labels"])

    title = plot_spec.get("title", "")
    subtitle = plot_spec.get("subtitle", "")
    axes.set_title(
        "\n".join(part for part in (title, subtitle) if part)
        or axes.get_title()
    )
    axes.set_xlabel(
        plot_spec.get("x_label") or data.get("x_label_hint") or "x"
    )
    axes.set_ylabel(plot_spec.get("y_label") or data.get("y_label_hint") or "y")
    if plot_spec.get("x_log"):
        axes.set_xscale("log")
    if plot_spec.get("y_log"):
        axes.set_yscale("log")
    if plot_spec.get("x_range"):
        axes.set_xlim(plot_spec["x_range"])
    if plot_spec.get("y_range"):
        axes.set_ylim(plot_spec["y_range"])
    apply_style(figure, axes, plot_spec, warnings)

    fmt = (plot_spec.get("output", {}) or {}).get("format", "svg")
    if fmt not in FORMATS:
        warnings.append(f"unsupported format {fmt!r}; rendered svg instead")
        fmt = "svg"
    output_path = output_path.with_suffix(f".{fmt}")
    output_path.parent.mkdir(parents=True, exist_ok=True)
    metadata = NO_DATE_METADATA.get(fmt, {})
    figure.savefig(
        output_path,
        format=fmt,
        dpi=dpi,
        metadata=metadata,
    )
    plt.close(figure)
    return {
        "schema_version": SCHEMA_VERSION,
        "backend": BACKEND,
        "plot_spec_sha256": canonical_hash(plot_spec),
        "artifact": {
            "path": output_path.name,
            "format": fmt,
            "width": width,
            "height": height,
            "dpi": dpi,
            "bytes": output_path.stat().st_size,
        },
        "data_summary": summary,
        "warnings": warnings,
    }


def main(argv: list[str]) -> int:
    request_path = result_path = None
    arguments = argv[1:]
    for index, argument in enumerate(arguments):
        if argument == "--request":
            request_path = Path(arguments[index + 1])
        elif argument == "--result":
            result_path = Path(arguments[index + 1])
    if request_path is None or result_path is None:
        print("usage: render.py --request REQUEST --result RESULT", file=sys.stderr)
        return 2
    request = json.loads(request_path.read_text(encoding="utf-8"))
    plot_spec = request["plot_spec"]
    output_path = request_path.parent / request.get("output_path", "figure.svg")
    try:
        result = render(plot_spec, output_path)
    except Exception as error:  # surfaced as harness failure envelope
        result = {
            "schema_version": SCHEMA_VERSION,
            "backend": BACKEND,
            "status": "error",
            "message": f"{type(error).__name__}: {error}",
        }
        result_path.write_text(
            json.dumps(result, indent=2, ensure_ascii=False) + "\n", encoding="utf-8"
        )
        return 1
    result_path.write_text(
        json.dumps(result, indent=2, ensure_ascii=False) + "\n", encoding="utf-8"
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))
